### PR TITLE
Python api for cpp model

### PR DESCRIPTION
**Pull request by zhenwei-intel**
## Type of Change

feature

## Description

detail description 
JIRA ticket: [NLPTOOLKIU-824](https://jira.devtools.intel.com/browse/NLPTOOLKIU-824)

- [x] Refact main to class Model
- [x] Pybind Model to python api
- [x] Refact postprocess
- [x] Add more parameters
- [x] Indentify token: end of text
- [x] pybind of quant model
- [x] init containing convertion and quantization
- [x] integrate python api with itrex.transformers AutoModel and generate
- [x] support weight only quant config

to be done in another pr
- convert script refactor
- More post processing
- Use transfomers tokenizer and streamer

## Expected Behavior & Potential Risk

```python
from intel_extension_for_transformers.transformers import AutoModelForCausalLM
prompt = "Once upon a time, a little girl"
model_name = "mosaicml/mpt-7b"
model = AutoModelForCausalLM.from_pretrained(model_name, use_llm_runtime=True)
print(model.generate(prompt))
```

or

```python
# 1. convert
from intel_extension_for_transformers.llm.runtime.graph.scripts.convert_model import convert_model
convert_model(model_name, "ne_mpt_f32.bin", "f32")

# 2. quantize
import intel_extension_for_transformers.llm.runtime.graph.Model as Model
Model.quant_model(model_path="ne_mpt_f32.bin", out_path="ne_mpt_q4_0.bin", bits=4)


# 3.inference
model = Model()
model.init_from_bin("mpt", "ne_mpt_q4_0.bin")

print(model.generate(prompt))
```

## Dependency Change?

pybind11 introduced

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/setup.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        output_dir = f"{extdir}{os.sep}"
+        cmake_args = [
+            f"-DCMAKE_LIBRARY_OUTPUT_DIRECTORY={output_dir}",
+            f"-DCMAKE_RUNTIME_OUTPUT_DIRECTORY={output_dir}",
+            f"-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{cfg.upper()}={output_dir}",
+            f"-DCMAKE_RUNTIME_OUTPUT_DIRECTORY_{cfg.upper()}={output_dir}",
+            f"-DPYTHON_EXECUTABLE={sys.executable}",
+            f"-DCMAKE_BUILD_TYPE={cfg}",  # not used on MSVC, but no harm
+        ]
+        if sys.platform == "linux":  # relative_rpath
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/setup.py:73]: please also consider windows part

File: intel_extension_for_transformers/llm/runtime/graph/setup.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        )
+        subprocess.run(
+            ["cmake", "--build", ".", *build_args], cwd=build_temp, check=True
+        )
+
+
+# The information here can also be placed in setup.cfg - better separation of
+# logic and declaration, and simpler if you include description/version in a file.
+setup(
+    name="itrex_llm",
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/setup.py:152]: itrex_llm_runtime ? or anyother option?

File: intel_extension_for_transformers/llm/runtime/graph/setup.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    version="1.0",
+    author="Intel AIA/AIPC Team",
+    description="Repository of cpp model in Intel® Intel Extension for Transformers",
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/setup.py:155]: Repository of llm runtime in Intel® Intel Extension for Transformers

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+#endif
+{
+    m.doc() = "cpp model python binding";
+    py::class_<Model>(m, "Model", py::module_local())
+        .def(py::init())
+        .def("init_model", &Model::init_model)
+        .def("generate", &Model::generate)
+        .def("set_n_predict", &Model::set_n_predict)
+        .def("set_batch_size", &Model::set_batch_size)
+        .def("set_ctx_size", &Model::set_ctx_size)
```
Comment (a32543254) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:274]: how about get rid of set 
like n_predict     batch_size  ? seems also ok

**Comment by a32543254** [comment]
please also add a main_llm.py to run a sample code of api 
like https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/deprecated/executor/python/test_model.py

File: intel_extension_for_transformers/llm/runtime/graph/setup.py (hunk 2 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/setup.py]: done

File: intel_extension_for_transformers/llm/runtime/graph/setup.py (hunk 3 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/setup.py]: done

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 4 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: updated like this:
```
m.init_model(model_path = "ne_mpt_q4_j.bin",
             n_predict = 100,
             seed = 12,
             threads = 28)
```

**Comment by zhenwei-intel** [comment]
> please also add a main_llm.py to run a sample code of api like https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/deprecated/executor/python/test_model.py

done, intel_extension_for_transformers/llm/runtime/graph/scripts/test_python_api.py

**Comment by airMeng** [comment]
can you give some comments @lvliang-intel  whether these API enough for chatbot application?

File: intel_extension_for_transformers/llm/runtime/graph/README.md [hunk 5]
```diff
@@ -115,7 +115,40 @@ LLM running script args explanations:
...
 
-### 4. Tensor Parallelism cross nodes/sockets
+Here is how to install itrex_llm_runtime from source.
+```bash
+git submodule update --recursive --init
+python setup.py install
+```
+
+```python
+import itrex_llm_runtime.mpt_cpp as cpp_model
```
Comment (hshen14) [inline intel_extension_for_transformers/llm/runtime/graph/README.md:127]: @a32543254 is there any better python api/package?

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/README.md:127]: we are working on this interface:

```
from intel_extension_for_transformers.transformers import AutoModel
model_name = "mosaicml/mpt-7b"
model = AutoModel.from_pretrained(model_name, use_cpp_graph=True)
model.gernerate(prompt = prompt, streamer)
```

File: intel_extension_for_transformers/llm/runtime/graph/scripts/python_api_example.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# from transformers import AutoTokenizer
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/scripts/python_api_example.py:1]: copyright?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+  candidates.reserve(n_vocab);
+  for (model_token token_id = 0; token_id < n_vocab; token_id++) {
+    candidates.emplace_back(model_token_data{token_id, logits[token_id], 0.0f});
+  }
+  model_token_data_array candidates_p = {candidates.data(), candidates.size(), false};
+
+  // Apply penalties
+  float nl_logit = logits[model_token_nl()];
+  auto last_n_repeat = std::min(std::min((int)last_n_tokens.size(), repeat_last_n), n_ctx);
+  model_sample_repetition_penalty(ctx, &candidates_p, last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:199]: chatbot only need this kind of post processing?

File: intel_extension_for_transformers/llm/runtime/graph/itrex_llm_runtime/convert_model.py [hunk 8]
```diff
@@ -22,6 +22,20 @@
...
+
+    path = Path(Path(__file__).parent.absolute(), "convert_{}.py".format(model_type))
+    cmd = []
+    cmd.extend(["python", path])
+    cmd.extend(["--outfile", outfile])
+    cmd.extend(["--outtype", outtype])
+    cmd.extend([model])
+
+    print("cmd:", cmd)
+    subprocess.run(cmd)
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/itrex_llm_runtime/convert_model.py:37]: avoid executing commands via sub-process, call these functions directly.

File: intel_extension_for_transformers/llm/runtime/graph/setup.py (hunk 1 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/setup.py]: remove this file

File: intel_extension_for_transformers/llm/runtime/graph/scripts/python_api_example.py (hunk 6 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/python_api_example.py]: updated

File: intel_extension_for_transformers/llm/runtime/graph/itrex_llm_runtime/convert_model.py (hunk 8 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/itrex_llm_runtime/convert_model.py]: Mark as TODO, I will refact convert script when I have enough time.

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 7 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:199]: More post processing will be add, maybe another pr?

Comment (airMeng) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:199]: how about leaving an option for post processing in advance? In this PR we only support ```--post-process (default)greedy-search```

**Comment by zhenwei-intel** [comment]
use itrex.transformers
```
from intel_extension_for_transformers.transformers import AutoModel, WeightOnlyQuantConfig
model_name = "/mnt/disk1/data2/zhenweil/models/mpt/mpt-7b"
woq_config = WeightOnlyQuantConfig(compute_dtype="ggml")
model = AutoModel.from_pretrained(model_name, quantization_config=woq_config, use_llm_runtime=True)
prompt = "Once upon a time, a little girl"
print(model.generate(prompt, sentence_mode=True, n_predict=30))
```

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 7 above)
Comment (a32543254) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:199]: will be fix in this pr

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:199]: arguments added, function will be refined.

File: intel_extension_for_transformers/llm/runtime/graph/README.md [hunk 9]
```diff
@@ -52,29 +52,67 @@ cmake ..
...
+### 2. Run LLM with Python API
+
+You can use the python api to simplely run HF model.
+```python
+from intel_extension_for_transformers.transformers import AutoModel, WeightOnlyQuantConfig
+model_name = "EleutherAI/gpt-j-6b"     # support model id of HF or local PATH to model
+woq_config = WeightOnlyQuantConfig(compute_dtype="int8", weight_dtype="int4")
+model = AutoModel.from_pretrained(model_name, quantization_config=woq_config, use_llm_runtime=True)
+prompt = "Once upon a time, a little girl"
+output = model.generate(prompt, streamer, max_new_tokens=30)
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/README.md:64]: does `streamer` class need initialize？

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+  float* logits = model_get_logits(ctx);
+  int next_token_id = post_process(logits);
+  curr_input_ids = {next_token_id};
+
+  if (next_token_id == model_token_eos() || n_past - prompt.size() == params.n_predict) {
+    token_eos = true;
+  }
+
+  auto next_token = model_token_to_str(ctx, next_token_id);
+  if (strcmp(next_token, "<|endoftext|>") == 0) {
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:131]: how about use `next_token_id == vocab.eos_token_id` for updating `token_eos` state?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+                          py::arg("post_process") = "topk"
+                          )
+      .def("generate", &Model::generate, "Generate tokens with prompt",
+                          py::arg("prompt"),
+                          py::arg("sentence_mode") = true
+                          )
+      .def_static("quant_model", &Model::quant_model, "Quantize model",
+                          py::arg("model_path"),
+                          py::arg("out_path"),
+                          py::arg("bits") = 4,
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:314]: so arg `bits` == arg `weight_dtype`? If so, why not change `bits` to `weight_dtype` for better understanding?

Comment (a32543254) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:314]: we will change this on pr https://github.com/intel/intel-extension-for-transformers/pull/320

File: intel_extension_for_transformers/llm/runtime/graph/README.md (hunk 9 above)
Comment (a32543254) [reply intel_extension_for_transformers/llm/runtime/graph/README.md:64]: this usage is same as transformers model.generate api
so we assume consumers know how to use streamers

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 10 above)
Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:131]: I check it here
```
  if (next_token_id == model_token_eos() || n_past - prompt.size() == params.n_predict) {
    token_eos = true;
  }
```
but some model will fail, so add endoftext checker

Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:131]: `model_token_eos` is for `llama.cpp` which return fixed value. We add each model's special tokens ids  including `eos` in `vocab` struct.

Comment (zhenwei-intel) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:131]: fixed, thx

**Comment by VincyZhang** [comment]
engine UT coverage drop due to scripts included into coverage calculation, omit those paths.
